spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala

```diff
@@ -0,0 +1,516 @@
+   */
+  protected def getFileChanges(
+    fromVersion: Long,
+    fromIndex: Long,
+    isStartingVersion: Boolean): Unit = {
+    if (!sortedFetchedFiles.isEmpty) { return }
+
+    if (fromVersion > intervalGetTableVersion) {
+      return
+    }
+
+    def sortAddFile(f1: AddFile, f2: AddFile): Boolean = { f1.url < f2.url }
+
+    def sortAddFileForCDF(f1: AddFileForCDF, f2: AddFileForCDF): Boolean = {
+      f1.version < f2.version || (f1.version == f2.version && f1.url < f2.url)
+    }
+
+    def appendToSortedFetchedFiles(indexedFile: IndexedFile): Unit = {
+      sortedFetchedFiles = sortedFetchedFiles :+ indexedFile
+    }
+
+    if (isStartingVersion) {
+      val tableFiles = deltaLog.client.getFiles(deltaLog.table, Nil, None, Some(fromVersion), None)
+      val numFiles = tableFiles.files.size
+      tableFiles.files.sortWith(sortAddFile).zipWithIndex.foreach{
+        case (file, index) if (index > fromIndex) =>
+          appendToSortedFetchedFiles(
+            IndexedFile(fromVersion, index, file, isLast = (index + 1 == numFiles)))
+      }
+    } else {
+      // TODO(lin.zhou) return metadata for each version, and check schema read compatibility
+      val tableFiles = deltaLog.client.getFiles(deltaLog.table, fromVersion)
+      val addFiles = verifyStreamHygieneAndFilterAddFiles(tableFiles)
+      addFiles.groupBy(a => a.version).toSeq.sortWith(_._1 < _._1).foreach{
+        case (v, adds) =>
+          val numFiles = adds.size
+          adds.sortWith(sortAddFileForCDF).zipWithIndex.foreach{
+            case (add, index) if (v > fromVersion || index > fromIndex) =>
+              appendToSortedFetchedFiles(IndexedFile(
+                add.version,
+                index,
+                AddFile(add.url, add.id, add.partitionValues, add.size, add.stats),
+                isLast = (index + 1 == numFiles))
+              )
+          }
+      }
+    }
+  }
+
+  protected def getLastFileChangeWithRateLimit(
+    fromVersion: Long,
+    fromIndex: Long,
+    isStartingVersion: Boolean,
+    limits: Option[AdmissionLimits] = Some(new AdmissionLimits())): Option[IndexedFile] = {
+    if (options.readChangeFeed) {
+      throw DeltaSharingErrors.CDFNotSupportedInStreaming
+    } else {
+      getFileChanges(fromVersion, fromIndex, isStartingVersion)
+
+      if (limits.isEmpty) return sortedFetchedFiles.lastOption
+
+      // Check each change until we've seen the configured number of addFiles. Return the last one
+      // for the caller to build offset.
+      var admissionControl = limits.get
+      var lastFileChange: Option[IndexedFile] = None
+      sortedFetchedFiles.foreach{indexedFile =>
+        if (admissionControl.admit(Option(indexedFile.add))) {
+          lastFileChange = Some(indexedFile)
+        } else {
+          return lastFileChange
+        }
+      }
+
+      lastFileChange
+    }
+  }
+
+  /**
+   * Get the changes from startVersion, startIndex to the endOffset, and create DataFrame
```

Comment:
Consider adding comments about side effects (how sortedFetchedFiles is manipulated).